cpp/tests/utilities/cxxopts.hpp

```diff
@@ -0,0 +1,1497 @@
+  void ensure_value(std::shared_ptr<const OptionDetails> details)
+  {
+    if (m_value == nullptr) { m_value = details->make_storage(); }
+  }
+
+  std::shared_ptr<Value> m_value;
+  size_t m_count = 0;
+  bool m_default = false;
+};
+
+class KeyValue {
+ public:
+  KeyValue(std::string key_, std::string value_)
+    : m_key(std::move(key_)), m_value(std::move(value_))
+  {
+  }
+
+  const std::string& key() const { return m_key; }
+
+  const std::string& value() const { return m_value; }
+
+  template <typename T>
+  T as() const
+  {
+    T result;
+    values::parse_value(m_value, result);
+    return result;
+  }
+
+ private:
+  std::string m_key;
+  std::string m_value;
+};
+
+class ParseResult {
+ public:
+  ParseResult(
+    const std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<OptionDetails>>>,
+    std::vector<std::string>,
+    bool allow_unrecognised,
+    int&,
+    char**&);
+
+  size_t count(const std::string& o) const
+  {
+    auto iter = m_options->find(o);
+    if (iter == m_options->end()) { return 0; }
+
+    auto riter = m_results.find(iter->second);
+
+    return riter->second.count();
+  }
+
+  const OptionValue& operator[](const std::string& option) const
+  {
+    auto iter = m_options->find(option);
+
+    if (iter == m_options->end()) { throw_or_mimic<option_not_present_exception>(option); }
+
+    auto riter = m_results.find(iter->second);
+
+    return riter->second;
+  }
+
+  const std::vector<KeyValue>& arguments() const { return m_sequential; }
+
+ private:
+  void parse(int& argc, char**& argv);
+
+  void add_to_option(const std::string& option, const std::string& arg);
+
+  bool consume_positional(std::string a);
+
+  void parse_option(std::shared_ptr<OptionDetails> value,
+                    const std::string& name,
+                    const std::string& arg = "");
+
+  void parse_default(std::shared_ptr<OptionDetails> details);
+
+  void checked_parse_arg(int argc,
+                         char* argv[],
+                         int& current,
+                         std::shared_ptr<OptionDetails> value,
+                         const std::string& name);
+
+  const std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<OptionDetails>>> m_options;
+  std::vector<std::string> m_positional;
+  std::vector<std::string>::iterator m_next_positional;
+  std::unordered_set<std::string> m_positional_set;
+  std::unordered_map<std::shared_ptr<OptionDetails>, OptionValue> m_results;
+
+  bool m_allow_unrecognised;
+
+  std::vector<KeyValue> m_sequential;
+};
+
+struct Option {
+  Option(const std::string& opts,
+         const std::string& desc,
+         const std::shared_ptr<const Value>& value = ::cxxopts::value<bool>(),
+         const std::string& arg_help               = "")
+    : opts_(opts), desc_(desc), value_(value), arg_help_(arg_help)
+  {
+  }
+
+  std::string opts_;
+  std::string desc_;
+  std::shared_ptr<const Value> value_;
+  std::string arg_help_;
+};
+
+class Options {
```

Comment:
Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`. Also, please use more descriptive names. Please provide a name that clearly differentiates between an `Option` abstraction and (I assume) a collection, `Options`. Just One "s" is not informative enough.